File: go/internal/tools/tools.go
package tools

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/lloydzhou/bash-agent/internal/config"
	"github.com/lloydzhou/bash-agent/internal/prompt"
	"github.com/lloydzhou/bash-agent/internal/safety"
)

type Result struct {
	Output string
	Err    error
}

type Runner struct {
	Config   config.Config
	TodoFile string
	Cwd      string
	Home     string
}

func (r Runner) Dispatch(name string, input json.RawMessage) Result {
	switch name {
	case "Read":
		var args struct {
			Path string `json:"path"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Read(args.Path)
		return Result{Output: out, Err: err}
	case "Write":
		var args struct {
			Path    string `json:"path"`
			Content string `json:"content"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Write(args.Path, args.Content)
		return Result{Output: out, Err: err}
	case "Edit":
		var args struct {
			Path      string `json:"path"`
			OldString string `json:"old_string"`
			NewString string `json:"new_string"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Edit(args.Path, args.OldString, args.NewString)
		return Result{Output: out, Err: err}
	case "Bash":
		var args struct {
			Command string `json:"command"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Bash(args.Command)
		return Result{Output: out, Err: err}
	case "Glob":
		var args struct {
			Pattern string `json:"pattern"`
			Path    string `json:"path"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Glob(args.Pattern, args.Path)
		return Result{Output: out, Err: err}
	case "Grep":
		var args struct {
			Pattern string `json:"pattern"`
			Path    string `json:"path"`
			Glob    string `json:"glob"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.Grep(args.Pattern, args.Path, args.Glob)
		return Result{Output: out, Err: err}
	case "TodoWrite":
		var args struct {
			Todos []struct {
				Content string `json:"content"`
				Status  string `json:"status"`
			} `json:"todos"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.TodoWrite(args.Todos)
		return Result{Output: out, Err: err}
	case "Skill":
		var args struct {
			Name string `json:"name"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.ToolSkill(args.Name)
		return Result{Output: out, Err: err}
	case "WebSearch":
		var args struct {
			Query string `json:"query"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.WebSearch(args.Query)
		return Result{Output: out, Err: err}
	case "WebFetch":
		var args struct {
			URL string `json:"url"`
		}
		if err := json.Unmarshal(input, &args); err != nil {
			return Result{Err: err}
		}
		out, err := r.WebFetch(args.URL)
		return Result{Output: out, Err: err}
	default:
		return Result{Err: fmt.Errorf("unknown tool: %s", name)}
	}
}

func (r Runner) Read(path string) (string, error) {
	if path == "" {
		return "", errors.New("Error: no path provided")
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("Error: file not found: %s", path)
		}
		if os.IsPermission(err) {
			return "", fmt.Errorf("Error: permission denied: %s", path)
		}
		return "", err
	}
	return string(data), nil
}

func (r Runner) Write(path, content string) (string, error) {
	if path == "" {
		return "", errors.New("Error: no path provided")
	}
	if len([]byte(content)) > r.Config.FileWriteMaxBytes {
		return "", fmt.Errorf("Error: content too large for write_file (%d bytes > %d bytes)", len([]byte(content)), r.Config.FileWriteMaxBytes)
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return "", err
	}
	// Preserve existing file permissions, default to 0o644 for new files
	perm := os.FileMode(0o644)
	if info, err := os.Stat(path); err == nil {
		perm = info.Mode().Perm()
	}
	if err := os.WriteFile(path, []byte(content), perm); err != nil {
		return "", err
	}
	info, err := os.Stat(path)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("OK: wrote %d bytes to %s", info.Size(), path), nil
}

func (r Runner) Edit(path, oldString, newString string) (string, error) {
	if path == "" {
		return "", errors.New("Error: no path provided")
	}
	if oldString == "" {
		return "", errors.New("Error: empty old_string")
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return "", fmt.Errorf("Error: file not found: %s", path)
		}
		return "", err
	}
	if len(data) > r.Config.FileWriteMaxBytes {
		return "", fmt.Errorf("Error: file too large for edit_file (%d bytes > %d bytes)", len(data), r.Config.FileWriteMaxBytes)
	}
	content := string(data)
	idx := strings.Index(content, oldString)
	if idx < 0 {
		return "", fmt.Errorf("Error: old_string not found in %s. Hint: Read the file and copy exact bytes (including whitespace/indent/newlines) before retrying Edit.", path)
	}
	updated := strings.Replace(content, oldString, newString, 1)
	if updated == "" {
		return "", errors.New("Error: edit produced empty result, reverted")
	}
	// Preserve original file permissions
	info, err := os.Stat(path)
	if err != nil {
		return "", fmt.Errorf("Error: cannot stat file: %s", path)
	}
	perm := info.Mode().Perm()
	diffOut, diffErr := unifiedDiffColor(path, content, updated)
	if err := os.WriteFile(path, []byte(updated), perm); err != nil {
		return "", err
	}
	if diffErr != nil {
		return "", diffErr
	}
	// Match bash tool_edit: output = summary_line + "\n" + colorized_diff + "\n"
	added, removed := countDiffLines(diffOut)
	summary := fmt.Sprintf("Edit(%s) [+%d -%d lines]", path, added, removed)
	if diffOut == "" {
		return fmt.Sprintf("Edit(%s) [no changes]", path), nil
	}
	return summary + "\n" + diffOut + "\n", nil
}

func unifiedDiffColor(path, oldContent, newContent string) (string, error) {
	oldFile, err := os.CreateTemp("", "edit-old-*")
	if err != nil {
		return "", err
	}
	defer os.Remove(oldFile.Name())
	defer oldFile.Close()
	newFile, err := os.CreateTemp("", "edit-new-*")
	if err != nil {
		return "", err
	}
	defer os.Remove(newFile.Name())
	defer newFile.Close()
	if _, err := oldFile.WriteString(oldContent); err != nil {
		return "", err
	}
	if _, err := newFile.WriteString(newContent); err != nil {
		return "", err
	}
	label := strings.TrimPrefix(path, "/")
	// Try with --color=always first (match bash tool_edit behavior)
	out, err := exec.Command("diff", "-u", "--color=always", "--label", "a/"+label, "--label", "b/"+label, oldFile.Name(), newFile.Name()).CombinedOutput()
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok && exitErr.ExitCode() == 1 {
			// diff returns 1 when differences found — check if --color was unsupported
			if strings.Contains(string(out), "unsupported --color") || strings.Contains(string(out), "unrecognized option '--color'") {
				// Fallback: no color
				out2, err2 := exec.Command("diff", "-u", "--label", "a/"+label, "--label", "b/"+label, oldFile.Name(), newFile.Name()).CombinedOutput()
				if err2 != nil {
					if exitErr2, ok := err2.(*exec.ExitError); ok && exitErr2.ExitCode() == 1 {
						return string(out2), nil
					}
					return "", fmt.Errorf("Error: diff failed")
				}
				return string(out2), nil
			}
			return string(out), nil
		}
		return "", fmt.Errorf("Error: diff failed")
	}
	return string(out), nil
}

// countDiffLines counts added/removed lines in unified diff output (works with ANSI color codes).
func countDiffLines(diff string) (added, removed int) {
	for _, line := range strings.Split(diff, "\n") {
		// Strip ANSI escape sequences for prefix check
		stripped := stripAnsi(line)
		if strings.HasPrefix(stripped, "+") && !strings.HasPrefix(stripped, "+++") {
			added++
		}
		if strings.HasPrefix(stripped, "-") && !strings.HasPrefix(stripped, "---") {
			removed++
		}
	}
	return
}

// stripAnsi removes ANSI escape sequences from a string.
func stripAnsi(s string) string {
	var out strings.Builder
	out.Grow(len(s))
	i := 0
	for i < len(s) {
		if s[i] == '\033' && i+1 < len(s) && s[i+1] == '[' {
			// Skip CSI sequence: ESC [ ... final_byte
			j := i + 2
			for j < len(s) && ((s[j] >= 0x30 && s[j] <= 0x3f) || (s[j] >= 0x20 && s[j] <= 0x2f)) {
				j++
			}
			if j < len(s) && s[j] >= 0x40 && s[j] <= 0x7e {
				j++
			}
			i = j
		} else {
			out.WriteByte(s[i])
			i++
		}
	}
	return out.String()
}

func (r Runner) Bash(command string) (string, error) {
	if command == "" {
		return "", errors.New("Error: no command provided")
	}
	if reason := safety.DenyBashCommandReason(command); reason != "" {
		return "", fmt.Errorf("Error: command blocked by bash safety policy (%s)", reason)
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(r.Config.ToolTimeoutSecs)*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "bash", "-lc", command)
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &out
	err := cmd.Run()
	if ctx.Err() == context.DeadlineExceeded {
		return out.String() + fmt.Sprintf("\n[... truncated, command timed out after %d seconds ...]", r.Config.ToolTimeoutSecs), nil
	}
	return out.String(), err
}

func (r Runner) Glob(pattern, path string) (string, error) {
	if pattern == "" {
		return "", errors.New("Error: no pattern provided")
	}
	if path == "" {
		path = "."
	}
	if info, err := os.Stat(path); err != nil || !info.IsDir() {
		return "", fmt.Errorf("Error: directory not found: %s", path)
	}
	if _, err := exec.LookPath("rg"); err != nil {
		return "", errors.New("Error: rg is required for glob")
	}
	cmd := exec.Command("rg", "--files", path, "-g", pattern)
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = ioDiscard{}
	_ = cmd.Run()
	return out.String(), nil
}

func (r Runner) Grep(pattern, path, glob string) (string, error) {
	if pattern == "" {
		return "", errors.New("Error: no pattern provided")
	}
	if path == "" {
		path = "."
	}
	if _, err := os.Stat(path); err != nil {
		return "", fmt.Errorf("Error: path not found: %s", path)
	}
	if _, err := exec.LookPath("rg"); err != nil {
		return "", errors.New("Error: rg is required for grep")
	}
	args := []string{"-n", "--color", "never"}
	if glob != "" {
		args = append(args, "--glob", glob)
	}
	args = append(args, "--", pattern, path)
	cmd := exec.Command("rg", args...)
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = ioDiscard{}
	_ = cmd.Run()
	return out.String(), nil
}

func (r Runner) TodoWrite(todos []struct {
	Content string `json:"content"`
	Status  string `json:"status"`
}) (string, error) {
	if r.TodoFile == "" {
		return "", errors.New("Error: todo file not configured")
	}
	var completed, inProgress int
	lines := make([]string, 0, len(todos))
	for _, todo := range todos {
		if todo.Content == "" {
			return "", errors.New("Error: todo item content is required")
		}
		switch todo.Status {
		case "pending":
			lines = append(lines, "- [ ] "+todo.Content)
		case "in_progress":
			inProgress++
			lines = append(lines, "- [ ] "+todo.Content)
		case "completed":
			completed++
			lines = append(lines, "- [x] "+todo.Content)
		default:
			return "", fmt.Errorf("Error: invalid todo status: %s", todo.Status)
		}
	}
	if inProgress > 1 {
		return "", errors.New("Error: todo_write allows at most one in_progress item")
	}
	checklist := strings.Join(lines, "\n")
	if err := os.WriteFile(r.TodoFile, []byte(checklist+"\n"), 0o644); err != nil {
		return "", err
	}
	return checklist, nil
}

func (r Runner) ToolSkill(name string) (string, error) {
	name = strings.TrimSpace(name)
	if name == "" {
		return "", errors.New("Error: no skill name provided")
	}
	skillFile := prompt.ResolveSkillFile(r.Cwd, r.Home, name)
	if skillFile == "" {
		return "", fmt.Errorf("Error: skill not found: %s", name)
	}
	data, err := os.ReadFile(skillFile)
	if err != nil {
		return "", err
	}
	baseDir := filepath.Dir(skillFile)
	content := strings.ReplaceAll(string(data), "${BASH_AGENT_SKILL_DIR}", baseDir)
	return fmt.Sprintf("Skill: %s\nBase directory: %s\n\n%s", name, baseDir, content), nil
}

func (r Runner) WebSearch(query string) (string, error) {
	if query == "" {
		return "", errors.New("Error: no query provided")
	}
	apiKey := os.Getenv("JINA_API_KEY")
	req, err := http.NewRequest("GET", "https://s.jina.ai/", nil)
	if err != nil {
		return "", fmt.Errorf("Error: creating search request: %w", err)
	}
	q := req.URL.Query()
	q.Set("q", query)
	req.URL.RawQuery = q.Encode()
	if apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+apiKey)
	}
	req.Header.Set("X-Respond-With", "no-content")
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	req = req.WithContext(ctx)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("Error: search request failed: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("Error: reading search response: %w", err)
	}
	return string(body), nil
}

func (r Runner) WebFetch(url string) (string, error) {
	if url == "" {
		return "", errors.New("Error: no url provided")
	}
	apiKey := os.Getenv("JINA_API_KEY")
	req, err := http.NewRequest("GET", "https://r.jina.ai/", nil)
	if err != nil {
		return "", fmt.Errorf("Error: creating fetch request: %w", err)
	}
	q := req.URL.Query()
	q.Set("url", url)
	req.URL.RawQuery = q.Encode()
	if apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+apiKey)
	}
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	req = req.WithContext(ctx)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("Error: fetch request failed: %w", err)
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("Error: reading fetch response: %w", err)
	}
	return string(body), nil
}

func FormatToolResult(s string, max int) string {
	b := []byte(s)
	if len(b) <= max {
		return s
	}
	size := len(b)
	marker := fmt.Sprintf("\n\n[... truncated: showing first/last portions of %d bytes ...]\n\n", size)
	markerLen := len(marker) + 20 // extra room for size digits
	tailLines := 5
	tailText := lastNLines(s, tailLines)
	tailLen := len(tailText)
	headLen := max - markerLen - tailLen
	if headLen <= 0 {
		headLen = max / 2
	}
	if headLen > size {
		headLen = size
	}
	headText := string(b[:headLen])
	return headText + marker + tailText
}

func lastNLines(s string, n int) string {
	count := 0
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == '\n' {
			count++
			if count >= n {
				return s[i+1:]
			}
		}
	}
	return s
}

type ioDiscard struct{}

func (ioDiscard) Write(p []byte) (int, error) { return len(p), nil }

File: pyagent/pyagent/tools.py
"""Tool implementations — async handlers, Bash uses asyncio subprocess."""
import asyncio
import fnmatch
import json
import os
import re

from pyagent.config import Config
from pyagent.util import format_tool_result


# ── helpers ──

def _param(params: dict, key: str, default=None):
    return params.get(key, default)


def _required(params: dict, key: str) -> str:
    v = params.get(key)
    if v is None:
        raise ValueError(f"Missing required parameter: {key}")
    return v


# ── Read ──

async def tool_read(params: dict, cfg: Config) -> str:
    path = _required(params, "path")
    offset = _param(params, "offset")
    limit = _param(params, "limit")

    if not os.path.isfile(path):
        return f"Error: file not found: {path}"

    try:
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            lines = f.readlines()
    except Exception as e:
        return f"Error reading file: {e}"

    if offset is not None:
        start = max(1, int(offset)) - 1
    else:
        start = 0

    if limit is not None:
        end = start + int(limit)
    else:
        end = len(lines)

    selected = lines[start:end]
    result = "".join(selected)
    return format_tool_result(result, cfg.tool_result_max_bytes)


# ── Write ──

async def tool_write(params: dict, cfg: Config) -> str:
    path = _required(params, "path")
    content = _required(params, "content")

    if len(content.encode("utf-8")) > cfg.file_write_max_bytes:
        return f"Error: content exceeds max write size ({cfg.file_write_max_bytes} bytes)"

    try:
        os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
        with open(path, "w", encoding="utf-8") as f:
            f.write(content)
    except Exception as e:
        return f"Error writing file: {e}"

    try:
        with open(path, "rb") as f:
            wrote_bytes = len(f.read())
    except Exception:
        wrote_bytes = "?"
    return f"OK: wrote {wrote_bytes} bytes to {path}"


# ── Edit ──

async def tool_edit(params: dict, cfg: Config) -> str:
    import tempfile

    path = _required(params, "path")
    old_string = _required(params, "old_string")
    new_string = _required(params, "new_string")

    if not os.path.isfile(path):
        return f"Error: file not found: {path}"

    try:
        with open(path, "r", encoding="utf-8", errors="replace") as f:
            content = f.read()
    except Exception as e:
        return f"Error reading file: {e}"

    # Check file size before editing (like bash-agent's edit_file.awk)
    file_size = len(content.encode("utf-8"))
    if file_size > cfg.file_write_max_bytes:
        return f"Error: file too large for edit_file ({file_size} bytes > {cfg.file_write_max_bytes} bytes)"

    count = content.count(old_string)
    if count == 0:
        return "Error: old_string not found in file"
    if count > 1:
        return f"Error: old_string found {count} times — must be unique"

    new_content = content.replace(old_string, new_string, 1)

    if len(new_content.encode("utf-8")) > cfg.file_write_max_bytes:
        return f"Error: content exceeds max write size ({cfg.file_write_max_bytes} bytes)"

    if not new_content:
        return "Error: edit produced empty result"

    # Write new content to temp file for diff
    tmp = tempfile.NamedTemporaryFile(mode="w", suffix=".tmp", delete=False, encoding="utf-8")
    try:
        tmp.write(new_content)
        tmp.close()

        # Run diff like bash-agent: diff -u --color=always --label a/path --label b/path
        label = path.lstrip("/")
        diff_output = ""
        try:
            proc = await asyncio.create_subprocess_exec(
                "diff", "-u", "--color=always",
                "--label", f"a/{label}", "--label", f"b/{label}",
                path, tmp.name,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=10)
            diff_output = stdout.decode("utf-8", errors="replace")
            stderr_text = stderr.decode("utf-8", errors="replace")
            # Fallback if --color=always not supported
            if "unsupported" in stderr_text or "unrecognized" in stderr_text:
                proc = await asyncio.create_subprocess_exec(
                    "diff", "-u",
                    "--label", f"a/{label}", "--label", f"b/{label}",
                    path, tmp.name,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                )
                stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=10)
                diff_output = stdout.decode("utf-8", errors="replace")
        except FileNotFoundError:
            diff_output = ""
        except asyncio.TimeoutError:
            diff_output = ""

        # Count added/removed lines (skip diff header lines starting with --- or +++ or @@)
        added = 0
        removed = 0
        for line in diff_output.split("\n"):
            if line.startswith("+") and not line.startswith("+++"):
                added += 1
            elif line.startswith("-") and not line.startswith("---"):
                removed += 1

        # Write final content
        with open(path, "w", encoding="utf-8") as f:
            f.write(new_content)

        result = f"Edit({path}) [+{added} -{removed} lines]"
        if diff_output:
            result += "\n" + diff_output
        return result
    finally:
        os.unlink(tmp.name)


# ── Bash (async subprocess) ──

async def tool_bash(params: dict, cfg: Config) -> str:
    command = _required(params, "command")
    timeout = _param(params, "timeout")
    if timeout is not None:
        timeout = int(timeout)

    from pyagent.safety import deny_bash_command_reason
    reason = deny_bash_command_reason(command)
    if reason:
        return f"Error: command blocked by bash safety policy ({reason})"

    try:
        proc = await asyncio.create_subprocess_exec(
            "bash", "-lc", command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=os.getcwd(),
        )
        stdout_bytes, stderr_bytes = await asyncio.wait_for(
            proc.communicate(),
            timeout=timeout or cfg.tool_timeout_secs,
        )
        output = ""
        if stdout_bytes:
            output += stdout_bytes.decode("utf-8", errors="replace")
        if stderr_bytes:
            if output:
                output += "\n"
            output += stderr_bytes.decode("utf-8", errors="replace")
        # bash-agent returns raw output, no exit code annotation
        return format_tool_result(output, cfg.tool_result_max_bytes)
    except asyncio.TimeoutError:
        timeout_val = timeout or cfg.tool_timeout_secs
        return f"[... command timed out after {timeout_val} seconds ...]"
    except Exception as e:
        return f"Error executing command: {e}"


# ── Glob ──

async def tool_glob(params: dict, cfg: Config) -> str:
    pattern = _required(params, "pattern")
    path = _param(params, "path", ".")

    if not os.path.isdir(path):
        return f"Error: directory not found: {path}"

    try:
        proc = await asyncio.create_subprocess_exec(
            "rg", "--files", path, "-g", pattern,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=30)
        output = stdout.decode("utf-8", errors="replace").strip()
        if not output:
            return "(no matches)"
        return output
    except FileNotFoundError:
        return "Error: rg is required for glob"
    except asyncio.TimeoutError:
        return "Error: glob timed out"
    except Exception as e:
        return f"Error: {e}"


# ── Grep ──

async def tool_grep(params: dict, cfg: Config) -> str:
    pattern = _required(params, "pattern")
    path = _param(params, "path", ".")
    context = _param(params, "context")
    glob_pattern = _param(params, "glob")

    if not os.path.exists(path):
        return f"Error: path not found: {path}"

    try:
        args = ["rg", "-n", "--color", "never"]
        if context and str(context).isdigit():
            args += ["-C", str(context)]
        if glob_pattern:
            args += ["--glob", glob_pattern]
        args += ["--", pattern, path]

        proc = await asyncio.create_subprocess_exec(
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=30)
        output = stdout.decode("utf-8", errors="replace").strip()
        if not output:
            return "(no matches)"
        return format_tool_result(output, cfg.tool_result_max_bytes)
    except FileNotFoundError:
        return "Error: rg is required for grep"
    except asyncio.TimeoutError:
        return "Error: grep timed out"
    except Exception as e:
        return f"Error: {e}"


# ── TodoWrite ──

def _parse_todos_to_markdown(todos: list) -> tuple[str, int, int, int]:
    """Parse todos array to markdown checklist, return (markdown, completed, total, in_progress)."""
    lines = []
    completed = 0
    total = 0
    in_progress = 0
    for item in todos:
        content = item.get("content", "")
        status = item.get("status", "pending")
        if not content:
            continue
        if status not in ("pending", "in_progress", "completed"):
            continue
        total += 1
        if status == "completed":
            completed += 1
            lines.append(f"- [x] {content}")
        elif status == "in_progress":
            in_progress += 1
            lines.append(f"- [>] {content}")
        else:
            lines.append(f"- [ ] {content}")
    return "\n".join(lines), completed, total, in_progress


async def tool_todo_write(params: dict, cfg: Config) -> str:
    todos = _required(params, "todos")

    if isinstance(todos, str):
        import json as _json
        try:
            todos = _json.loads(todos)
        except _json.JSONDecodeError:
            return "Error: todos must be a JSON array"

    if not isinstance(todos, list):
        return "Error: todos must be a JSON array"

    markdown, completed, total, in_progress = _parse_todos_to_markdown(todos)

    try:
        with open(cfg.todo_file, "w", encoding="utf-8") as f:
            f.write(markdown + "\n")
    except Exception as e:
        return f"Error writing todo file: {e}"

    # Store summary for display
    cfg._todo_summary = f"{completed}/{total}"

    # Emit todo_update event
    if cfg.log_events and cfg.session_event_file:
        from pyagent.session import session_append_line
        import json as _json
        session_append_line(cfg, _json.dumps({"type": "todo_update", "content": markdown}))

    # Stream-json: emit todo_update event
    if cfg.output_format == "stream-json":
        from pyagent.util import json_escape
        import sys
        sys.stdout.write(f'{{"type":"todo_update","content":"{json_escape(markdown)}"}}\n')
        sys.stdout.flush()

    return markdown


# ── Skill ──

def _find_skill_base_dirs() -> list[str]:
    """Match bash-agent's find_skill_base_dirs: cwd/.claude/skills, cwd/skills, ~/.claude/skills."""
    cwd = os.getcwd()
    home = os.path.expanduser("~")
    dirs = []
    for d in [
        os.path.join(cwd, ".claude", "skills"),
        os.path.join(cwd, "skills"),
        os.path.join(home, ".claude", "skills"),
    ]:
        if os.path.isdir(d):
            dirs.append(d)
    return dirs


async def tool_skill(params: dict, cfg: Config) -> str:
    name = _required(params, "name").strip()
    if not name:
        return "Error: no skill name provided"

    for base_dir in _find_skill_base_dirs():
        skill_dir = os.path.join(base_dir, name)
        skill_file = os.path.join(skill_dir, "SKILL.md")
        if os.path.isfile(skill_file):
            with open(skill_file, "r", encoding="utf-8", errors="replace") as f:
                content = f.read()
            # Replace ${BASH_AGENT_SKILL_DIR} with actual directory
            content = content.replace("${BASH_AGENT_SKILL_DIR}", skill_dir)
            skill_content = f"Base directory: {skill_dir}\n\n{content}"
            return f"Skill: {name}\n{skill_content}"

    return f"Error: skill not found: {name}"


# ── WebSearch (Jina API) ──

async def tool_web_search(params: dict, cfg: Config) -> str:
    import aiohttp

    query = _required(params, "query")
    jina_key = os.environ.get("JINA_API_KEY", "")

    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(
                "https://s.jina.ai/",
                params={"q": query},
                headers={
                    "Authorization": f"Bearer {jina_key}",
                    "X-Respond-With": "no-content",
                },
                timeout=aiohttp.ClientTimeout(total=30, connect=10),
            ) as resp:
                return await resp.text()
    except Exception as e:
        return f"Error: web search failed: {e}"


# ── WebFetch (Jina API) ──

async def tool_web_fetch(params: dict, cfg: Config) -> str:
    import aiohttp

    url = _required(params, "url")
    jina_key = os.environ.get("JINA_API_KEY", "")

    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(
                "https://r.jina.ai/",
                params={"url": url},
                headers={"Authorization": f"Bearer {jina_key}"},
                timeout=aiohttp.ClientTimeout(total=60, connect=10),
            ) as resp:
                return await resp.text()
    except Exception as e:
        return f"Error: web fetch failed: {e}"


# ── Dispatch ──

TOOL_DISPATCH = {
    "Read": tool_read,
    "Write": tool_write,
    "Edit": tool_edit,
    "Bash": tool_bash,
    "Glob": tool_glob,
    "Grep": tool_grep,
    "TodoWrite": tool_todo_write,
    "Skill": tool_skill,
    "WebSearch": tool_web_search,
    "WebFetch": tool_web_fetch,
}


async def dispatch_tool(name: str, params: dict, cfg: Config) -> str:
    """Execute a tool by name, return output string."""
    handler = TOOL_DISPATCH.get(name)
    if not handler:
        return f"Error: unknown tool: {name}"
    try:
        return await handler(params, cfg)
    except Exception as e:
        return f"Error: tool execution failed: {e}"
